pulsar-client-kafka-compat/pulsar-client-kafka/src/main/java/org/apache/kafka/clients/producer/PulsarKafkaProducer.java

```diff
@@ -0,0 +1,247 @@
+import org.apache.pulsar.client.kafka.compat.MessageIdUtils;
+import org.apache.pulsar.client.kafka.compat.PulsarKafkaConfig;
+
+public class PulsarKafkaProducer<K, V> implements Producer<K, V> {
+
+    private final PulsarClient client;
+    private final ProducerConfiguration pulsarProducerConf;
+
+    private final ConcurrentMap<String, org.apache.pulsar.client.api.Producer> producers = new ConcurrentHashMap<>();
+
+    private final Serializer<K> keySerializer;
+    private final Serializer<V> valueSerializer;
+
+    /** Map that contains the last future for each producer */
+    private final ConcurrentMap<String, CompletableFuture<MessageId>> lastSendFuture = new ConcurrentHashMap<>();
+
+    public PulsarKafkaProducer(Map<String, Object> configs) {
+        this(configs, null, null);
+    }
+
+    public PulsarKafkaProducer(Map<String, Object> configs, Serializer<K> keySerializer,
+            Serializer<V> valueSerializer) {
+        this(configs, new Properties(), keySerializer, valueSerializer);
+    }
+
+    public PulsarKafkaProducer(Properties properties) {
+        this(properties, null, null);
+    }
+
+    public PulsarKafkaProducer(Properties properties, Serializer<K> keySerializer, Serializer<V> valueSerializer) {
+        this(new HashMap<>(), properties, keySerializer, valueSerializer);
+    }
+
+    @SuppressWarnings({ "unchecked", "deprecation" })
+    private PulsarKafkaProducer(Map<String, Object> conf, Properties properties, Serializer<K> keySerializer,
+            Serializer<V> valueSerializer) {
+        properties.forEach((k, v) -> conf.put((String) k, v));
+
+        ProducerConfig producerConfig = new ProducerConfig(conf);
+
+        if (keySerializer == null) {
+            this.keySerializer = producerConfig.getConfiguredInstance(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG,
+                    Serializer.class);
+            this.keySerializer.configure(producerConfig.originals(), true);
+        } else {
+            this.keySerializer = keySerializer;
+            producerConfig.ignore(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG);
+        }
+
+        if (valueSerializer == null) {
+            this.valueSerializer = producerConfig.getConfiguredInstance(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG,
+                    Serializer.class);
+            this.valueSerializer.configure(producerConfig.originals(), true);
+        } else {
+            this.valueSerializer = valueSerializer;
+            producerConfig.ignore(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG);
+        }
+
+        String serviceUrl = producerConfig.getList(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG).get(0);
+        ClientConfiguration clientConf = PulsarKafkaConfig.getClientConfiguration(properties);
+        try {
+            client = PulsarClient.create(serviceUrl, clientConf);
+        } catch (PulsarClientException e) {
+            throw new RuntimeException(e);
+        }
+
+        pulsarProducerConf = new ProducerConfiguration();
```

Comment:
how about setting `compression.type` ? Pulsar and Kafka both support `lz4 and gzip`

Reply:
Good point, we can easily do that.